File: Programmers/타겟 넘버.swift
import Foundation

func solution(_ numbers:[Int], _ target:Int) -> Int {
    var answer = 0
    
    func calculate(depth: Int, target: Int, numbers: [Int], value: Int) {
        //탈출조건
        if depth == numbers.count {
            if value == target {
                answer += 1
            }
            return
        }
        
        calculate(depth: depth+1, target: target, numbers: numbers, value: value+numbers[depth])
        calculate(depth: depth+1, target: target, numbers: numbers, value: value-numbers[depth])
    }
    
    calculate(depth: 1, target: target, numbers: numbers, value: numbers[0])
    calculate(depth: 1, target: target, numbers: numbers, value: -numbers[0])
    
    return answer
}

print(solution([1,1,1,1,1], 3))

File: BOJ/타겟넘버.py
def recursive(numbers, index, target, sum, result):
    if index < len(numbers):
        recursive(numbers, index+1, target, sum + numbers[index], result)
        recursive(numbers, index+1, target, sum + -(numbers[index]), result)
    else:
        if sum == target:
            result.append(1)

def solution(numbers, target):
    answer = []
    recursive(numbers, 0, target, 0, answer)
    return sum(answer)


print(solution([4, 1, 2, 1], 4))